src/pricecypher/rest.py

```diff
@@ -0,0 +1,257 @@
+from random import randint
+from datetime import datetime
+from marshmallow import Schema
+
+from .exceptions import PriceCypherError, RateLimitError
+
+UNKNOWN_ERROR = 'pricecypher.sdk.internal.unknown'
+
+
+class RestClientOptions(object):
+    """
+    Configuration object for RestClient.
+    Used for configuring additional RestClient options, such as rate-limit retries.
+
+    :param float or tuple[float,float] timeout: (optional) Change the requests connect and read timeout (in seconds).
+        Pass a tuple to specify both values separately or a float to set both to it.
+        (defaults to 300.0 (5 minutes) for both)
+    :param int retries: (optional) In the event an API request returns a 429 response header (indicating rate-limit
+        has been hit), the RestClient will retry the request this many times using an exponential backoff strategy,
+        before raising a RateLimitError exception.
+        (defaults to 3)
+    """
+
+    def __init__(self, timeout=None, retries=None):
+        self.timeout = 300.0
+        self.retries = 3
+
+        if timeout is not None:
+            self.timeout = timeout
+
+        if retries is not None:
+            self.retries = retries
+
+
+class RestClient(object):
+    """
+    Provides simple methods for handling all RESTful api endpoints.
+
+    :param str jwt: JWT token used to authorize requests to the APIs.
+    :param RestClientOptions options: (optional) Pass an instance of RestClientOptions to configure additional
+        RestClient options, such as rate-limit retries.
+    """
+
+    def __init__(self, jwt, options=None):
+        if options is None:
+            options = RestClientOptions()
+
+        self.options = options
+        self.jwt = jwt
+
+        self._metrics = {'retries': 0, 'backoff': []}
+        self._skip_sleep = False
+
+        self.base_headers = {
+            'Authorization': 'Bearer {}'.format(self.jwt),
+            'Content-Type': 'application/json',
+            'Accept': 'application/json',
+        }
+
+    # Returns the maximum amount of jitter to introduce in milliseconds (100ms)
+    def MAX_REQUEST_RETRY_JITTER(self):
+        return 100
+
+    # Returns the maximum delay window allowed (1000ms)
+    def MAX_REQUEST_RETRY_DELAY(self):
+        return 1000
+
+    # Returns the minimum delay window allowed (100ms)
+    def MIN_REQUEST_RETRY_DELAY(self):
+        return 100
+
+    def _retry(self, make_request):
+        # Track the API request attempt number
+        attempt = 0
+
+        # Reset the metrics tracker
+        self._metrics = {'retries': 0, 'backoff': []}
+
+        # Floor the retries at 0.
+        retries = max(0, self.options.retries)
+
+        while True:
+            # Increment attempt number
+            attempt += 1
+
+            # Issue the request
+            response = make_request()
+
+            # break iff no retry needed
+            if response.status_code != 429 or retries <= 0 or attempt > retries:
```

Comment:
```suggestion
            if response.status_code != 429 or attempt > retries:
```

Reply:
Won't this introduce the possibility for an infinite loop when retries (which is passed by the user of this class) is negative?

Reply:
No, then attempt will be larger then retries for sure (as attempt is always 1+). Either way, the retries parameter is maxed with 0 so sending negative values doesn't make a difference. But, fine to keep for clarity, just not really needed to see this edge case as an edge case.